### ISSUE TITLE
Fix S3800 (`function-return-type`): add TS to compatible languages

### ISSUE THREAD

**Issue by ilia-kebets-sonarsource** [SonarSource/SonarJS#4060]
Add "TYPESCRIPT" here:

https://github.com/SonarSource/SonarJS/blob/c9f8688c01198425ff1436e3bda3f76ab4a7e756/sonar-plugin/javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3800.json#L17-L19

**Comment by yassin-kammoun-sonarsource** [comment]
I am not sure it makes sense to enable this rule for TypeScript since it supports union types. WDYT?

**Comment by ilia-kebets-sonarsource** [comment]
I mean, it's allowed, just as it is allowed in JS, but we want to enforce the single-type return, no?

**Comment by victor-diez-sonarsource** [comment]
not sure neither myself, pretty common to have union types `| undefined` or `| null`, unless we modify the rule to have those as exceptions. wdyt? @yassin-kammoun-sonarsource @ilia-kebets-sonarsource

**Comment by yassin-kammoun-sonarsource** [comment]
I still think we should not enable this rule for TypeScript. Looking at what we do ourselves in SonarJS, we do use them, and it would be weird to return an object with two possible values instead of a union as a fix for an issue if we really need to return values from different types.

**Comment by victor-diez-sonarsource** [comment]
Also, you will be able to bypass this rule by just creating a union type:

```ts
type ReturnType = ReturnType1 | ReturnType2 | ReturnType3

function f(): ReturnType {
   if (cond1) {
      return new ReturnType1();
   }
   else if (cond2) {
      return new ReturnType2();
   }
   return new ReturnType3();
}
```

@ilia-kebets-sonarsource WDYT? @gabriel-vivas-sonarsource ?

**Comment by ilia-kebets-sonarsource** [comment]
Couldn't we identify that the return type is a union since we're in TS?

**Comment by victor-diez-sonarsource** [comment]
Closed as don't planned. TypeScript already doing the job and this rule would just create noise in most cases

**Comment by ilia-kebets-sonarsource** [comment]
this was not done
